File: test/arkHelpersTest.py
import sys, os

sys.path.append(os.path.abspath('../ieUtil/'))
import arkHelpers

import ieInit
ieInit.init()

import unittest

class arkHelpersTest(unittest.TestCase):

	def setUp(self):
		pass

	def tearDown(self):
		pass

	def test_pad(self):
		self.assertEqual(arkHelpers.pad(1, 4), '0001')
		self.assertEqual(arkHelpers.pad(15, 7), '0000015')
		self.assertEqual(arkHelpers.pad(131, 2), '131')

	def test_varType(self):
		self.assertEqual(arkHelpers.varType(4), 'int')
		self.assertEqual(arkHelpers.varType('string'), 'str')
		self.assertEqual(arkHelpers.varType([1,2,3]), 'list')

	# def test_parseJSON(self):
	# 	json = '["foo", {"bar":["baz", null, 1.0, 2]}]'
	# 	parsed = arkHelpers.parseJSON(json, True)
	# 	print parsed
	# 	self.assertEqual(parsed[0], 'Foo')

	def test_postStrings(self):
		self.assertEqual(arkHelpers.postString({'tea': 'pot', 'cat' : 'dog', 'wow': 'amaze'}), 'tea=pot&wow=amaze&cat=dog')

	def test_mergeDict(self):
		self.assertEqual(arkHelpers.mergeDict({'a': 'b', 'c': 'd'} , {'e': 'f'}), {'a': 'b', 'c': 'd', 'e': 'f'})
		self.assertEqual(arkHelpers.mergeDict({1: 2}, {}), {1: 2})

	def test_movieSafeDim(self):
		self.assertEqual(arkHelpers.movieSafeDim(4), 4)
		self.assertEqual(arkHelpers.movieSafeDim(3.9), 0)
		self.assertEqual(arkHelpers.movieSafeDim(4.1), 4)
		self.assertEqual(arkHelpers.movieSafeDim(11), 8)

	def test_safeFilename(self):
		self.assertEqual(arkHelpers.safeFilename('mike.txt'), 'mike.txt')
		self.assertEqual(arkHelpers.safeFilename('mike^^^.txt'), 'mike.txt')
		self.assertEqual(arkHelpers.safeFilename('   &&&mike.txt^%&__'), 'mike.txt__')

	def test_randomHash(self):
		self.assertEqual(len(arkHelpers.randomHash()), 16)
		self.assertEqual(len(arkHelpers.randomHash(4)), 4)

	def test_makeArrayUnique(self):
		self.assertEqual(arkHelpers.makeArrayUnique([1,2,4]), [1,2,4])
		self.assertEqual(arkHelpers.makeArrayUnique([1,2,3,4,3]), [1,2,3,4])
		self.assertEqual(arkHelpers.makeArrayUnique([1,2,4,4,4,4,4,4]), [1,2,4])
		self.assertEqual(arkHelpers.makeArrayUnique([1,2,4,'a','b','c',1,2,4,'a','b','c']), [1,2,4, 'a','b','c'])

	def test_makeWebSafe(self):
		self.assertEqual(arkHelpers.makeWebSafe('mike%is%cool'), 'mike_is_cool')
		self.assertEqual(arkHelpers.makeWebSafe('mike1234'), 'mike1234')
		self.assertEqual(arkHelpers.makeWebSafe('mike__1234'), 'mike__1234')
		self.assertEqual(arkHelpers.makeWebSafe('    mike  1234^^  '), '____mike__1234____')

	def test_getExtension(self):
		self.assertEqual(arkHelpers.getExtension('mike.txt'), '.txt')
		self.assertEqual(arkHelpers.getExtension('/path/to/file.yolo'), '.yolo')
		self.assertEqual(arkHelpers.getExtension('path/to/file'), '')

	def test_normalizeExtension(self):
		self.assertEqual(arkHelpers.normalizeExtension('.TXT'), '.txt')
		self.assertEqual(arkHelpers.normalizeExtension('png'), '.png')
		self.assertEqual(arkHelpers.normalizeExtension('Yolo'), '.yolo')

	def test_removeExtension(self):
		self.assertEqual(arkHelpers.removeExtension('mike.txt'), 'mike')
		self.assertEqual(arkHelpers.removeExtension('path/to/file.psd'), 'path/to/file')
		self.assertEqual(arkHelpers.removeExtension('mike'), 'mike')

	def test_ensureExtension(self):
		self.assertEqual(arkHelpers.ensureExtension('Mike.txt', '.txt'), 'Mike.txt')
		self.assertEqual(arkHelpers.ensureExtension('Mike', '.txt'), 'Mike.txt')
		self.assertEqual(arkHelpers.ensureExtension('Mike.txt', '.mb'), 'Mike.txt.mb')

	def test_parseCommaArray(self):
		self.assertEqual(arkHelpers.parseCommaArray('Mike, is, cool'), ['Mike', 'is', 'cool'])
		self.assertEqual(arkHelpers.parseCommaArray('like, comments'), ['like', 'comments'])
		self.assertEqual(arkHelpers.parseCommaArray('           wow           ,           amaze            '), ['wow', 'amaze'])

	def test_appendOrSetArray(self):
		self.assertEqual(arkHelpers.appendOrSetArray([1,2,3], 4), [1,2,3,4])
		self.assertEqual(arkHelpers.appendOrSetArray(1, [1,2,3]), [1,2,3])
		self.assertEqual(arkHelpers.appendOrSetArray(1, 3), [3])
		self.assertEqual(arkHelpers.appendOrSetArray(None, None), [])

	def test_ensureArray(self):
		self.assertEqual(arkHelpers.ensureArray([1,2,3]), [1,2,3])
		self.assertEqual(arkHelpers.ensureArray(1), [1])
		self.assertEqual(arkHelpers.ensureArray(None), [])

	def test_ensureNumber(self):
		self.assertEqual(arkHelpers.ensureNumber(4), 4)
		self.assertEqual(arkHelpers.ensureNumber(189.3), 189.3)
		self.assertEqual(arkHelpers.ensureNumber('a'), 0)

	def test_omitObjectKeys(self):
		self.assertEqual(arkHelpers.omitObjectKeys({1:2, 3:4, 5:6}, [1]), {3:4, 5:6})
		self.assertEqual(arkHelpers.omitObjectKeys({1:2, 3:4, 5:6}, [3,5]), {1: 2})
		self.assertEqual(arkHelpers.omitObjectKeys({1:2, 3:4, 5:6}, [4]), {1:2, 3:4, 5:6})

	def test_collectObjectKeys(self):
		self.assertEqual(arkHelpers.collectObjectKeys({1:2, 3:4, 5:6}, [3, 5]), {3:4, 5:6})
		self.assertEqual(arkHelpers.collectObjectKeys({1:2, 3:4, 5:6}, [1]), {1: 2})
		self.assertEqual(arkHelpers.collectObjectKeys({1:2, 3:4, 5:6}, [4]), {})

	def test_parseInt(self):
		self.assertEqual(arkHelpers.parseInt(4), 4)
		self.assertEqual(arkHelpers.parseInt('4a'), 4)
		self.assertEqual(arkHelpers.parseInt('4'), 4)
		self.assertEqual(arkHelpers.parseInt('abc'), 0)

	def test_capitalize(self):
		self.assertEqual(arkHelpers.capitalize('mike'), 'Mike')
		self.assertEqual(arkHelpers.capitalize('Mike'), 'Mike')
		self.assertEqual(arkHelpers.capitalize('MIKE'), 'MIKE')
		self.assertEqual(arkHelpers.capitalize(' mike'), ' mike')
		self.assertEqual(arkHelpers.capitalize('7mike'), '7mike')

	def test_capitalizeWords(self):
		self.assertEqual(arkHelpers.capitalizeWords('mike is great'), 'Mike Is Great')
		self.assertEqual(arkHelpers.capitalizeWords('   wow so nice'), '   Wow So Nice')
		self.assertEqual(arkHelpers.capitalizeWords(' 7lo 5m2'), ' 7lo 5m2')

	def test_parseSort(self):
		self.assertEqual(arkHelpers.parseSort("field:ASC", None), {'field' : 'field', 'order': 'ASC', 'combined' : 'field:ASC'})
		self.assertEqual(arkHelpers.parseSort("field", 1), {'field' : 'field', 'order': 'ASC', 'combined' : 'field:ASC'})
		self.assertEqual(arkHelpers.parseSort("field:descending", None), {'field' : 'field', 'order': 'DSC', 'combined' : 'field:DSC'})

	def test_stringCompare(self):
		self.assertEqual(arkHelpers.stringCompare('a', 'b'), -1)
		self.assertEqual(arkHelpers.stringCompare('not mike', 'mike'), 1)
		self.assertEqual(arkHelpers.stringCompare('wow', 'wow'), 0)

	def test_getAlphaNumericOnly(self):
		self.assertEqual(arkHelpers.getAlphaNumericOnly('mike'), 'mike')
		self.assertEqual(arkHelpers.getAlphaNumericOnly('mike__123'), 'mike123')
		self.assertEqual(arkHelpers.getAlphaNumericOnly('^$%@^#'), '')

if __name__ == '__main__':
	unittest.main()

File: test/arkHelpersTest.js
// Vendor Modules
////////////////////////
// var _ = require('lodash')
var expect = require('expect.js')

var describe = global.describe
var it = global.it


describe('helpers', function() {

	var helpers

	it ('should load', function() {
		helpers = require('../arkHelpers/arkUtil.js')
	})

	it ('should pad', function() {
		expect(helpers.pad(42, 4)).to.be('0042')
		expect(helpers.pad(5, 6)).to.be('000005')
		expect(helpers.pad(123, 2)).to.be('123')
	})

	it ('should varType', function() {
		expect(helpers.varType(4)).to.be('number')
		expect(helpers.varType({1: 2})).to.be('object')
		expect(helpers.varType('abc')).to.be('string')
	})

	it ('should randomHash', function() {
		expect(helpers.randomHash().length).to.be(16)
		expect(helpers.randomHash(4).length).to.be(4)
	})

	it ('should makeArrayUnique', function() {
		expect(helpers.makeArrayUnique([1,2,3]).length).to.be([1,2,3].length)
		expect(helpers.makeArrayUnique([1,2,3,4,3]).length).to.be([1,2,3,4].length)
		expect(helpers.makeArrayUnique(['a','b','c','a','b','c']).length).to.be(['a','b','c'].length)
		expect(helpers.makeArrayUnique([1,'a','a',1]).length).to.be([1, 'a'].length)
	})

	it ('should safeFilename', function() {
		expect(helpers.safeFilename('m4#1%')).to.equal('m41')
		expect(helpers.safeFilename('mikke')).to.be('mikke')
		expect(helpers.safeFilename('&&&test!!!')).to.be('test')
	})

	it ('should movieSafeDim', function() {
		expect(helpers.movieSafeDim(4)).to.be(4)
		expect(helpers.movieSafeDim(7)).to.be(4)
		expect(helpers.movieSafeDim(8.1)).to.be(8)
		expect(helpers.movieSafeDim(11.99999)).to.be(8)
	})

	it ('should makeWebSafe', function() {
		expect(helpers.makeWebSafe('mike%is%cool')).to.be('mike_is_cool')
		expect(helpers.makeWebSafe('website')).to.be('website')
		expect(helpers.makeWebSafe('WEBSITE^^^IN^CAPS')).to.be('website___in_caps')
	})

	it ('should postString', function() {
		expect(helpers.postString({'a': 'b', 'cat': 'dog'})).to.be('a=b&cat=dog')
		expect(helpers.postString({123: 456, 'yolo': 'swag'})).to.be('123=456&yolo=swag')
	})

	it ('should mergeObject', function() {
		expect(String(helpers.mergeObject({1: 2}, {2: 3}))).to.equal(String({1: 2, 2: 3}))
		expect(String(helpers.mergeObject({'a': 'b', 1: 2}, {'c': 'd', 2: 1, 1: 3}))).to.equal(String({'a': 'b', 1: 3, 'c': 'd', 2: 1}))
	})

	it ('should parseCommaArray', function() {
		expect(helpers.parseCommaArray('like, comments')).to.contain('like')
		expect(helpers.parseCommaArray('like, comments')).to.contain('comments')
		expect(helpers.parseCommaArray('   wow   , so   , nice')).to.contain('wow')
		expect(helpers.parseCommaArray('   wow   , so   , nice')).to.contain('so')
		expect(helpers.parseCommaArray('   wow   , so   , nice')).to.contain('nice')
		expect(helpers.parseCommaArray('')).to.contain('')
	})

	it ('should appendOrSetArray', function() {
		expect(helpers.appendOrSetArray([1,2,3], 4)).to.contain(4)
		expect(helpers.appendOrSetArray(1, 3)).to.contain(3)
		expect(helpers.appendOrSetArray(1, [3])).to.contain(3)
		expect(helpers.appendOrSetArray(null, null).length).to.be(0)
	})

	it ('should ensureArray', function() {
		expect(helpers.ensureArray([1,2,3])).to.contain(1)
		expect(helpers.ensureArray([1,2,3])).to.contain(2)
		expect(helpers.ensureArray([1,2,3])).to.contain(3)
		expect(helpers.ensureArray(1)).to.contain(1)
		expect(helpers.ensureArray().length).to.be(0)
	})

	it ('should ensureNumber', function() {
		expect(helpers.ensureNumber(4)).to.be(4)
		expect(helpers.ensureNumber('a')).to.be(0)
		expect(helpers.ensureNumber('3a')).to.be(3)
		expect(helpers.ensureNumber('34')).to.be(34)
		expect(helpers.ensureNumber('4.2')).to.be(4.2)
	})

	it ('should omitObjectKeys', function() {
		var obj = {1: 2, 3: 4, 5: 6}
		expect(helpers.omitObjectKeys(obj, [1])).to.eql({3: 4, 5: 6})
		expect(helpers.omitObjectKeys(obj, [3, 5])).to.eql({1: 2})
		expect(helpers.omitObjectKeys(obj, 3)).to.eql({1: 2, 5: 6})
		expect(helpers.omitObjectKeys(obj, [4])).to.eql(obj)
	})

	it ('should collectObjectKeys', function() {
		var obj = {1: 2, 3: 4, 5: 6}
		expect(helpers.collectObjectKeys(obj, [1])).to.eql({1: 2})
		expect(helpers.collectObjectKeys(obj, [3, 5])).to.eql({3: 4, 5: 6})
		expect(helpers.collectObjectKeys(obj, 3)).to.eql({3: 4})
		expect(helpers.collectObjectKeys(obj, [4])).to.eql({})
	})

	it ('should parseJSON', function() {
		var obj = '{ "name": "Mike" }'
		var parsed = helpers.parseJSON(obj)
		expect(parsed.name).to.be("Mike")
	})

	it ('should parseSort', function() {
		expect(helpers.parseSort('field:ASC')).to.eql({'field': 'field', 'order': 'ASC', 'combined': 'field:ASC'})
		expect(helpers.parseSort('field:1')).to.eql({'field': 'field', 'order': 'ASC', 'combined': 'field:ASC'})
		expect(helpers.parseSort('field:descending')).to.eql({'field': 'field', 'order': 'DSC', 'combined': 'field:DSC'})
	})

	it ('should parseInt', function() {
		expect(helpers.parseInt(4)).to.be(4)
		expect(helpers.parseInt('3')).to.be(3)
		expect(helpers.parseInt('4b')).to.be(4)
		expect(helpers.parseInt('yolo')).to.be(0)
	})

	it ('should capitalize', function() {
		expect(helpers.capitalize('mike')).to.be('Mike')
		expect(helpers.capitalize('MIKE')).to.be('MIKE')
	})

	it ('should capitalizeWords', function() {
		expect(helpers.capitalizeWords('test this function')).to.be('Test This Function')
		expect(helpers.capitalizeWords('kevin van zonneveld')).to.be('Kevin Van Zonneveld')
		expect(helpers.capitalizeWords('   lots   OF .  spaCes')).to.be('   Lots   OF .  SpaCes')
	})

	it ('should stringCompare', function() {
		expect(helpers.stringCompare('a', 'b')).to.be(-1)
		expect(helpers.stringCompare('not mike', 'mike')).to.be(1)
		expect(helpers.stringCompare('oh hey', 'oh hey')).to.be(0)
	})

	it ('should joinURL', function() {
		expect(helpers.joinURL('url1', 'url2')).to.be('url1url2')
		expect(helpers.joinURL('url1', 54, 'url2')).to.be('url1url2')
	})

	it ('should getAlphaNumericOnly', function() {
		expect(helpers.getAlphaNumericOnly('Mike')).to.be('Mike')
		expect(helpers.getAlphaNumericOnly('Mike%%%1234')).to.be('Mike1234')
		expect(helpers.getAlphaNumericOnly('***Mike Wow***  ')).to.be('MikeWow')
	})

})